sid/visualize_simulation_results.py

```diff
@@ -2,249 +2,227 @@
-    for groupby_var in background_vars:
-        gb_title = f"Infection Related Rates by {_nice_str(groupby_var)}"
-        gb_title = Div(text=gb_title, style=style, name=f"div_{groupby_var}_inf_title")
-        gb_rates = _plot_rates_by_group(
-            data=data,
-            groupby_var=groupby_var,
-            infection_vars=infection_vars,
-            colors=get_colors("categorical", 12),
-        )
-        elements += [gb_title, *gb_rates]
-
-        # r zeros
-        r_title = f"Basic Replication Rate Overall and by {_nice_str(groupby_var)}"
-        r_title = Div(text=r_title, style=style, name="div_r_title")
-        r_zeros = _plot_r_zeros(data=data, groupby_var=groupby_var)
-        elements += [r_title, r_zeros]
-    return elements
-
-
-def _plot_infection_rates(data, infection_vars, colors):
-    overall_gb = data.groupby("date")
-    overall_means = overall_gb[infection_vars].mean()
-    title = "Infection Related Rates in the General Population"
-    p = _plot_rates(means=overall_means, colors=colors, title=title)
-    p.name = "plot_incidences/general"
-    return p
+def _create_statistics(data_paths, infection_vars, background_vars, window_length):
+    """Calculate the infection rates and reproduction numbers for each date.
 
+    Args:
+        data_paths (list): list of paths to the pickled simulation results
+        infection_vars (list): list of infection rates to plot
+        background_vars (list): list of background variables by whose value to group
+            the results. Have to be present in all simulation results.
+        window_length (int): how many dates to use for the reproduction numbers.
+
+    Returns:
+        rates (pd.DataFrame): DataFrame with the dates as index.
+            The columns are a MultiIndex with four levels: The outermost is the
+            "bg_var" ("general" for the overall rate).
+            The next is the "rate" (e.g. the infectious rate or r zero),
+            then "bg_value", the value of the background variable and last "data_id".
 
-def _plot_rates_by_group(data, groupby_var, infection_vars, colors):
-    gb = data.groupby(["date", groupby_var])
-    plots = []
-    if is_categorical(data[groupby_var]):
-        ordered = data[groupby_var].cat.ordered
-        n_categories = len(data[groupby_var].cat.categories)
-    else:
-        ordered = False
-    for var in infection_vars:
-        plot_colors = get_colors(f"blue-red", n_categories) if ordered else colors
-        means = gb[var].mean().unstack()
-        title = f"{_nice_str(var)} Rates by {_nice_str(groupby_var)}"
-        p = _plot_rates(means=means, colors=plot_colors, title=title)
-        p.name = f"plot_incidences/{groupby_var}/{var}_rate"
-        plots.append(p)
-    return plots
+    """
+    vars_for_r_zero = ["immune", "infection_counter", "cd_infectious_false"]
+    keep_vars = sorted(set(infection_vars + background_vars + vars_for_r_zero))
+    name_to_means = {}
+    for path in data_paths:
```

Comment:
Is the calculations of statistics between different data sets independent from each other? If so, I would extract this loop, because you are calling the "real" function within itself for each data set. The function becomes purer.

Reply:
You are right that I can extract the loop. I have never been quite satisfied with this function with its nestedness and its not-quite-repetitiveness. Your suggestion makes it quite a bit nicer. Thanks!